data_prep/introspector.py

```diff
@@ -613,8 +613,34 @@ def _combine_functions(a: list[str], b: list[str], c: list[str],
   return list(combined)
 
 
+def _select_functions_from_jvm_oracles(project: str, limit: int,
+                                       target_oracles: list[str]) -> list[dict]:
+  """Selects functions from oracles designated for jvm projects, with
+  jvm-public-candidates as the prioritised oracle"""
+  all_functions = OrderedDict()
+
+  if 'jvm-public-candidates' in target_oracles:
+    # JPC is the primary oracle for JVM projects. If it does exist, all other
+    # oracles are ignored because the results from all other oracles are subsets
+    # of the results from JPC oracle for JVM projects.
+    jpc_targets = _select_top_functions_from_oracle(project, limit,
+                                                    'jvm-public-candidates',
+                                                    target_oracles)
+    all_functions.update(jpc_targets)
+  else:
+    # JPC does not exist in target_oracles, follow given oracle order and
+    # dedulicate instead.
+    for target_oracle in target_oracles:
+      tmp_functions = _select_top_functions_from_oracle(project, limit,
+                                                        target_oracle,
+                                                        target_oracles)
+      all_functions.update(tmp_functions)
+
+  return list(all_functions.values())[:limit]
```

Comment:
nit, avoid repeated code:
```python
if 'jvm-public-candidates' in target_oracles:
  target_oracles = ['jvm-public-candidates']
for target_oracle in target_oracles:
      tmp_functions = _select_top_functions_from_oracle(project, limit,
                                                        target_oracle,
                                                        target_oracles)
      all_functions.update(tmp_functions)
  return list(all_functions.values())[:limit]
```
A cleaner way is to enforce this when [parsing args](https://github.com/google/oss-fuzz-gen/blob/9c4780d8c743beb86b19b578a3703a32bddc8623/data_prep/introspector.py#L909). E.g.:
```python
args = parser.parse_args()
if 'jvm-public-candidates' in args.target_oracles:
  args.target_oracles = ['jvm-public-candidates']
```
or even
```python
args = parser.parse_args()
assert ('jvm-public-candidates' not in args.target_oracles or args.target_oracles == ['jvm-public-candidates'], 'JPC is the primary oracle for JVM projects...'
```

Reply:
I though the first way is better, cause it allows better flexibility when we don't want jvm-public-candidates in some situation and when we want to test some other oracles solely.

Reply:
Please correct me if I am wrong, but I thought all four snippets (yours + my three in the comment) were functionally equivalent: They all remove other oracles if and only if `'jvm-public-candidates'` is in `target_oracles`.

The main difference is _where_ to implement this logic. I suggested the last one (`assert`) so that our users will be reminded that other oracles will not be used if `jvm-public-candidates` exists. Similarly, the second last way intends to prepare the **correct** oracle list as early as possible so that the **wrong** list won't confuse us/users.

If `jvm-public-candidates` is not in the oracle list, then all four snippets should behave the same.